feat(#85): expose selectedIndexes to tooltip renderer

canastro opened the pull request:
This is a proposal to https://github.com/entronad/graphic/issues/85 and it's a breaking change:

### Context

In our app we are using a `tooltipRenderer` but we need to be able to customize the display of each field on the tooltip based on information we dont have available in the tuples that are provided to Graphic. Therefore we need to get back the "index" of the selected tuple so that I can get the appropriate record from our data structure. Thats the gist of what we need. 

To give a bit more context, we're dynamically creating charts in a spreadsheet application with data provided by the user and, as google sheets we want to allow the user to have values with different formatting in each point. 

Here is an example where the user defined different number of decimal places per cell via formatting, ie. the user did not introduce 1.000000, the actual value is "1" but the formatting is "use 6 decimal places => 1.000000":


https://user-images.githubusercontent.com/1267623/163705323-2cd0d570-ce5c-448b-aeba-9c06f8d7c8ac.mov

## Proposal
In the tooltipRenderer return back the "selectedIndexes" to allow the user to find the appropriate record on a more complex datastructure on user-land. For users not using "tooltipRenderer", provide "tooltipFieldFormatter" which receives the selectedIndex for the current tuple being formatted.

Let me know what you think of the overall approach and if there are any alternative ways of doing this. 

Thank you.

canastro commented on example/lib/pages/custom_formatters.dart:
```diff
@@ -0,0 +1,221 @@
+                    )
+                  },
+                  tooltip: TooltipGuide(
+                    tooltipFieldFormatter: ({
+                      required field,
+                      required scale,
+                      required selectedIndex,
+                      required selectedTuple,
+                    }) {
+                      if (field == 'price') {
+                        return _kData[selectedIndex].formattedPrice;
+                      }
+
+                      return scale.format(selectedTuple[field]) ?? 'what?';
+                    },
```
Example of the tooltipFieldFormatter

canastro commented on example/lib/pages/custom_formatters.dart:
```diff
@@ -0,0 +1,221 @@
+  }
+}
+
+List<Figure> simpleTooltip({
+  required Offset anchor,
+  required List<Tuple> selectedTuples,
+  required Set<int> selectedIndexes,
+}) {
+  List<Figure> figures;
+
+  final data = _kData[selectedIndexes.first];
+  final textContent = 'Day: ${data.day}\n'
+      'Price: ${data.formattedPrice}\n'
+      'Stock ${data.stock} ';
```
Example of the tooltip renderer using the `selectedIndexes`

canastro commented on example/lib/pages/custom_formatters.dart:
```diff
@@ -0,0 +1,221 @@
+                      line: StrokeStyle(color: const Color(0xFF000000)),
+                      label: LabelStyle(
+                          offset: const Offset(0, 7.5),
+                          span: (value) {
+                            final parsedValue =
+                                double.tryParse(value)?.toStringAsFixed(0) ??
+                                    value;
+                            return TextSpan(
+                              text: 'day $parsedValue',
+                              style: const TextStyle(
+                                color: Color(0xFF000000),
+                                fontSize: 10,
+                              ),
+                            );
```
I'll probably need the index here as well, so I can check my tuples for the formatted value

entronad commented:
1. To add the selectedIndexes is a good idea. With it, you can find your original datum, and make usage of it's special attributes which are not in the tuples. There are some conflicts of this PR, so for convenience I will add it myself.
2. I don't think "tooltipFieldFormatter" is a good idea. It's function are all covered by the renderer, and it's hard to get what this argument really means. (I got the meaning of "field formatter" only after I carefully read the source code.). So, with the scale.formatter for common users, and renderer for senior users, I think the tooltipFieldFormatter is not that necessary.

canastro commented:
@entronad I misread the part `for convenience I will add it myself.` and rebased and simplified this PR. 
Feel free to ignore this PR and close it if you're already working on it.

canastro commented on example/lib/pages/polygon_custom.dart:
```diff
@@ -48,10 +48,11 @@ class TriangleShape extends IntervalShape {
-  Offset anchor,
-  List<Tuple> selectedTuples,
-) {
+List<Figure> simpleTooltip({
+  required Offset anchor,
+  required List<Tuple> selectedTuples,
+  required Set<int> selectedIndexes,
+}) {
```
I haven't created any example for the selectedIndexes, let me know if you want me to add one.

entronad commented:
Added in v0.10.0.
